### ISSUE TITLE
Issue 1: Refactor the experimental pipeline into reusable and reproducible code

### ISSUE THREAD

**Issue by Mati86** [AI-Security-Internships-2026/12-graph-based-cyber-physical-risk#22]
# Issue 1: Refactor the experimental pipeline into reusable and reproducible code

## Milestone: M1 — Reproducible Evaluation Foundation 

---

## Problem Statement

Convert the current notebook-heavy experimental workflow into a small, reusable research codebase so that all paper experiments can be re-run with the same preprocessing, train/validation/test splits, seeds, model settings, and result-export format.

The goal is **not** to rewrite the entire repository or build production software. The goal is to make the research experiments reproducible and comparable.

---

## Background

Most of the current implementation lives inside the weekly notebooks. This was acceptable during exploratory work, but the paper now requires controlled comparisons between:

- different split strategies,
- non-GNN and GNN baselines,
- topology/content ablations,
- temporal protocols,
- multiple random seeds.

If each experiment has slightly different preprocessing or evaluation code, the comparisons will not be scientifically reliable.

---

### 1. Reusable Module Structure

Create or update a compact reusable structure under `src/`. A suggested structure is:

```text
src/
├── data/
│   ├── scadanet.py
│   ├── batadal.py
│   └── ics_flow.py
├── graph/
│   ├── construction.py
│   └── rewiring.py
├── models/
│   ├── tabular.py
│   └── gnn.py
├── evaluation/
│   ├── metrics.py
│   ├── splits.py
│   ├── audit.py
│   └── temporal.py
├── experiments/
│   └── runner.py
└── utils/
    ├── seed.py
    └── io.py
```

This exact layout is not mandatory. A simpler clean design is acceptable.

### 2. Common Experiment Interface

Implement a common experiment interface capable of selecting:

- dataset,
- model,
- split protocol,
- random seed,
- threshold,
- result output path.

Example expected usage:

```bash
python -m src.experiments.runner \
  --dataset scadanet \
  --model graphsage \
  --split track_b \
  --seed 42 \
  --output experiments/results/q1/scadanet_graphsage_trackb_seed42.json
```

### 3. Strict Reproducibility Controls

Also support strict reproducibility controls:

- Python seed,
- NumPy seed,
- PyTorch seed,
- CUDA seed where applicable,
- deterministic split generation,
- saved split indices or IDs,
- configuration saved with every result.

Do not rely on a notebook kernel's previous state.

### 4. Reproduce Reference Experiments

Reproduce at least the following **existing reference experiments** using the refactored pipeline:

1. SCADANet Track B GraphSAGE.
2. SCADANet current temporal split.
3. BATADAL current temporal split.
4. One existing static BATADAL experiment.

The purpose is not to obtain new headline results in this issue. The purpose is to prove that the refactored pipeline reproduces the currently accepted values within normal seed/run variation.

### 5. Result Artifacts and Schema

Create a new directory such as:

```text
experiments/results/q1/
```

Every result file should record at minimum:

```text
dataset
model
split_protocol
seed
train_size
validation_size
test_size
feature_set
graph_mode
threshold
accuracy
precision
recall
f1
auprc
fpr
confusion_matrix
runtime
library_versions
git_commit
```

If a field is not applicable, explicitly store `null` or document why.

Also create:

```text
experiments/results/q1/reproducibility_check.csv
```

Suggested columns:

| experiment_id | old_result | reproduced_result | absolute_difference | status |
|---|---:|---:|---:|---|

#### Table R1 — Reproducibility of inherited experiments

| Dataset | Protocol | Metric | Previously reported | Refactored pipeline | Difference |
|---|---|---|---:|---:|---:|
| SCADANet | Track B | F1 | ... | ... | ... |
| SCADANet | Temporal | F1 | ... | ... | ... |
| BATADAL | Temporal | F1 | ... | ... | ... |
| BATADAL | Static CV | F1 | ... | ... | ... |

This table may appear in supplementary material rather than the main paper.

### 7. Optional Repository Figure

**Figure R1 — Reproducible experiment workflow**

```text
Dataset
  ↓
Preprocessing
  ↓
Saved split
  ↓
Model configuration
  ↓
Training
  ↓
Evaluation
  ↓
JSON/CSV result
```

This can later be reused as part of the methodology figure. No major paper figure is mandatory for this issue.

---

## Acceptance Criteria

- [ ] Core preprocessing/training/evaluation logic is moved from notebooks into reusable modules.
- [ ] Existing notebooks can import the reusable code instead of containing duplicated core implementations.
- [ ] All random seeds are explicitly controlled.
- [ ] Split definitions can be saved and reused exactly.
- [ ] At least four inherited experiments are reproduced successfully.
- [ ] Result JSON contains model, data, split, seed, metrics, threshold, and environment metadata.
- [ ] A single command can run an experiment without manually executing previous notebook cells.
- [ ] Existing paper results are not silently changed; differences are documented.
- [ ] Obsolete duplicate code may be removed only after confirming that the new implementation reproduces the intended experiment.

---

## Do Not Do In This Issue

Do not:

- design a new GNN architecture,
- add a dashboard,
- add new datasets,
- rewrite the research paper,
- optimize model performance,
- introduce unrelated features.

This issue is complete when the research workflow is reproducible.
